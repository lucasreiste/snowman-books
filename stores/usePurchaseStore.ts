import { defineStore } from "pinia";
import { ref } from "vue";
import type { Book } from "@/types/book";

interface FormData {
  name: string;
  email: string;
  address: string;
  cardNumber?: string;
  expiry?: string;
  cvv?: string;
}

export const usePurchaseStore = defineStore("purchase", () => {
  // Armazena os dados do livro
  const book = ref<Book | null>(null);

  // Armazena dados do formulário
  const formData = ref<FormData>({
    name: "",
    email: "",
    address: "",
    cardNumber: "",
    expiry: "",
    cvv: "",
  });

  // Ações (setters e reset)
  function setBook(newBook: Book) {
    book.value = newBook;
  }

  function resetFormData() {
    formData.value = {
      name: "",
      email: "",
      address: "",
      cardNumber: "",
      expiry: "",
      cvv: "",
    };
  }

  return {
    book,
    formData,
    setBook,
    resetFormData,
  };
});
