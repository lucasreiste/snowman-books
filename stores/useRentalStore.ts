import { defineStore } from "pinia";
import type { Book } from "@/types/book";
import type { RentalPersonalData, RentalDetails } from "@/types/rental";

interface RentalState {
  currentStep: number;
  selectedBook: Book | null;
  personalData: RentalPersonalData;
  rentalDetails: RentalDetails;
}

export const useRentalStore = defineStore("rental", {
  state: (): RentalState => ({
    currentStep: 1,
    selectedBook: null,
    personalData: {
      name: "",
      email: "",
      address: "",
    },
    rentalDetails: {
      duration: 7,
      totalPrice: 0,
    },
  }),

  actions: {
    setSelectedBook(book: Book) {
      this.selectedBook = book;
    },

    setPersonalData(data: Partial<RentalState["personalData"]>) {
      this.personalData = { ...this.personalData, ...data };
    },

    setRentalDetails(details: Partial<RentalState["rentalDetails"]>) {
      this.rentalDetails = { ...this.rentalDetails, ...details };
    },

    nextStep() {
      this.currentStep++;
    },

    previousStep() {
      this.currentStep--;
    },

    resetState() {
      this.$reset();
    },
  },
});
