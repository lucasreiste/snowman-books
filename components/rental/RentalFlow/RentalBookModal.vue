<template>
  <Dialog :open="open" @close="handleClose">
    <DialogContent class="lg:max-w-2xl p-6 [&>button]:hidden">
      <div class="flex flex-col">
        <RentalStepper :steps="steps" :current-step="rentalStore.currentStep" />
        <component :is="currentStepComponent" />
        <div class="mt-6 flex justify-end space-x-4">
          <Button
            v-if="rentalStore.currentStep === 1"
            variant="secondary"
            @click="handleClose"
          >
            {{ t("rental.buttons.close") }}
          </Button>
          <Button
            v-if="
              rentalStore.currentStep > 1 &&
              rentalStore.currentStep < steps.length
            "
            variant="secondary"
            @click="rentalStore.previousStep"
          >
            {{ t("rental.buttons.back") }}
          </Button>
          <Button
            v-if="rentalStore.currentStep < steps.length"
            variant="default"
            :disabled="isStepInvalid"
            @click="handleNextStep"
          >
            {{ t("rental.buttons.continue") }}
          </Button>
          <Button
            v-if="rentalStore.currentStep === steps.length"
            variant="default"
            @click="handleConfirm"
          >
            {{ t("rental.buttons.close") }}
          </Button>
        </div>
      </div>
    </DialogContent>
  </Dialog>
</template>

<script setup lang="ts">
import { computed, watch } from "vue";
import { useRental } from "@/composables/useRental";
import { useRentalStore } from "@/stores/useRentalStore";
import RentalConfirmationStep from "./steps/RentalConfirmationStep.vue";
import RentalPersonalDataStep from "./steps/RentalPersonalDataStep.vue";
import RentalPaymentStep from "./steps/RentalPaymentStep.vue";
import RentalSuccessStep from "./steps/RentalSuccessStep.vue";
import { useI18n } from "vue-i18n";

interface Step {
  step: number;
  title: string;
  description: string;
}

const props = defineProps<{
  open: boolean;
}>();

const emit = defineEmits<{
  (e: "close" | "confirm"): void;
}>();

const rentalStore = useRentalStore();
const rental = useRental();
const { t } = useI18n();

const steps: Step[] = [
  {
    step: 1,
    title: t("rental.steps.confirmation.title"),
    description: t("rental.steps.confirmation.description"),
  },
  {
    step: 2,
    title: t("rental.steps.personalData.title"),
    description: t("rental.steps.personalData.description"),
  },
  {
    step: 3,
    title: t("rental.steps.payment.title"),
    description: t("rental.steps.payment.description"),
  },
  {
    step: 4,
    title: t("rental.steps.success.title"),
    description: t("rental.steps.success.description"),
  },
];

const stepComponents: {
  [key: number]:
    | typeof RentalConfirmationStep
    | typeof RentalPersonalDataStep
    | typeof RentalPaymentStep
    | typeof RentalSuccessStep;
} = {
  1: RentalConfirmationStep,
  2: RentalPersonalDataStep,
  3: RentalPaymentStep,
  4: RentalSuccessStep,
};

const currentStepComponent = computed(
  () => stepComponents[rentalStore.currentStep]
);

const isStepInvalid = computed(() => {
  if (rentalStore.currentStep === 1 && !rentalStore.rentalDetails.totalPrice) {
    return true;
  }
  if (
    rentalStore.currentStep === 2 &&
    !rental.validatePersonalData(rentalStore.personalData)
  ) {
    return true;
  }
  return false;
});

async function handleNextStep() {
  rentalStore.nextStep();
}

async function handleConfirm() {
  emit("confirm");
  handleClose();
}

function handleClose() {
  emit("close");
}

watch(
  () => props.open,
  (isOpen) => {
    if (!isOpen) {
      rentalStore.resetFormState();
    }
  }
);
</script>
