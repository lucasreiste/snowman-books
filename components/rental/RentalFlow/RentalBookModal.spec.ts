/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, fireEvent } from "@testing-library/vue";
import { nextTick } from "vue";
import { createTestingPinia } from "@pinia/testing";
import RentalBookModal from "./RentalBookModal.vue";
import { useRentalStore } from "@/stores/useRentalStore";

const DialogMock = {
  name: "Dialog",
  template: '<div v-if="open" data-testid="modal"><slot /></div>',
  props: ["open"],
  emits: ["close"],
};

const DialogContentMock = {
  template: '<div data-testid="modal-content"><slot /></div>',
};

const ButtonMock = {
  template: `
    <button 
      :data-testid="$attrs['data-testid']" 
      :disabled="disabled" 
      @click="$emit('click')"
    >
      <slot />
    </button>
  `,
  props: ["disabled", "variant"],
  emits: ["click"],
  inheritAttrs: true,
};

vi.mock("@/composables/useRental", () => ({
  useRental: () => ({
    validatePersonalData: (data: any) => Object.keys(data).length > 0,
  }),
}));

describe("RentalBookModal", () => {
  let pinia: any;

  const renderComponent = (props = {}) => {
    pinia = createTestingPinia({
      createSpy: vi.fn,
      initialState: {
        rental: {
          currentStep: 1,
          rentalDetails: {
            totalPrice: 0,
          },
          personalData: {},
        },
      },
    });

    return render(RentalBookModal, {
      props: {
        open: true,
        ...props,
      },
      global: {
        plugins: [pinia],
        stubs: {
          Dialog: DialogMock,
          DialogContent: DialogContentMock,
          Button: ButtonMock,
          RentalStepper: true,
          RentalConfirmationStep: true,
          RentalPersonalDataStep: true,
          RentalPaymentStep: true,
          RentalSuccessStep: true,
        },
      },
    });
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("should render correctly when open", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it("should not render when closed", () => {
    renderComponent({ open: false });
    expect(screen.queryByTestId("modal")).not.toBeInTheDocument();
  });

  it("should emit close event when close button is clicked", async () => {
    const { emitted } = renderComponent();
    const closeButton = screen.getByText("Fechar");
    await fireEvent.click(closeButton);
    expect(emitted().close).toBeTruthy();
  });

  it("should disable continue button on first step when total price is not set", () => {
    const store = useRentalStore();
    store.rentalDetails.totalPrice = 0;
    renderComponent();

    const continueButton = screen.getByText("Continuar");
    expect(continueButton).toBeDisabled();
  });

  it("should enable continue button on first step when total price is set", async () => {
    renderComponent();
    const store = useRentalStore(pinia);
    store.rentalDetails.totalPrice = 100;

    await nextTick();
    const continueButton = screen.getByText("Continuar");
    expect(continueButton).not.toBeDisabled();
  });

  it("should show back button after first step", async () => {
    renderComponent();
    const store = useRentalStore(pinia);
    store.currentStep = 2;

    await nextTick();
    expect(screen.getByText("Voltar")).toBeInTheDocument();
  });

  it("should move to next step when continue is clicked", async () => {
    const store = useRentalStore();
    store.rentalDetails.totalPrice = 100;
    renderComponent();

    const continueButton = screen.getByText("Continuar");
    await fireEvent.click(continueButton);
    expect(store.currentStep).toBe(2);
  });
});
