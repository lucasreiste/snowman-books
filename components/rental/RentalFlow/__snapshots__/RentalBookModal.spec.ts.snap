// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RentalBookModal > should render correctly when open 1`] = `
<div>
  <div
    data-testid="modal"
  >
    
    <div
      class="lg:max-w-2xl p-6"
      data-testid="modal-content"
    >
      
      <div
        class="flex flex-col"
      >
        <rental-stepper-stub
          current-step="1"
          steps="[object Object],[object Object],[object Object],[object Object]"
        />
        <rental-confirmation-step-stub />
        <div
          class="mt-6 flex justify-end space-x-4"
        >
          <button>
            
            Fechar
            
          </button>
          <!--v-if-->
          <button
            disabled=""
          >
            
            Continuar
            
          </button>
          <!--v-if-->
        </div>
      </div>
      
    </div>
    
  </div>
</div>
`;
